dashboard/app/controllers/omniauth_callbacks_controller.rb

```diff
@@ -213,18 +211,19 @@ def sign_up_clever
     user = User.from_omniauth(auth_hash, auth_params, session)
     prepare_locale_cookie user
 
-    if user.persisted?
-      handle_untrusted_email_signin(user, AuthenticationOption::CLEVER)
-    elsif (looked_up_user = User.find_by_email_or_hashed_email(user.email))
-      email_already_taken_redirect(
-        provider: AuthenticationOption::CLEVER,
-        found_provider: looked_up_user.provider,
-        email: user.email
-      )
-    else
-      # This is a new registration
-      register_new_user user
-    end
+    # if the registration credentials identify us as an existing user, simply
+    # sign in as that user.
+    return sign_in_user user if user.persisted?
+
+    # if a user account with a conflicting email (that could not be
+    # authenticated with the given registration credentals) exists, display an
+    # interim page prompting the user to either log in to that account or
+    # create a new one with a manually-provided email.
+    email_conflict = User.find_by_email_or_hashed_email(user.email).present?
```

Comment:
I had to think for a bit about how this interacts with https://github.com/code-dot-org/code-dot-org/pull/31948 - my understanding is that now we _could_ remove this because technically our validation wouldn't catch a conflict and we could proceed to create the user as-is.  However, we are leaving this flow in to try and minimize such email conflicts and give teachers an early chance to link Clever to their existing account.  Do I have that right?

Reply:
Precisely! In fact, I should rebrand this; it's not so much about preventing a potential email conflict as it is about proactively detecting an existing matching account.